drivers: adc: atmel_sam: Add DMA support

mcscholtz opened the pull request:
This patch adds DMA support for the atmel sam s70/e70/v70/v71.
This patch requires that a hardware ADC trigger be configured.
The only hardware trigger that was tested was the timer counter.

Tested on a custom sams70 board.

Signed-off-by: Marius Scholtz <mariuss@ricelectronics.com>

mbolivar-nordic commented on drivers/adc/adc_sam_afec.c:
```diff
@@ -279,6 +341,50 @@ static int adc_sam_init(const struct device *dev)
+	data->timer = NULL;
+
+	/* If a timer is the trigger source, enable it now */
+	if (cfg->trigger_source == 1 ||
+	    cfg->trigger_source == 2 ||
+	    cfg->trigger_source == 3) {
+		switch (cfg->tc) {
+		case 0:
+			data->timer = device_get_binding("TC0");
+			break;
+		case 1:
+			data->timer = device_get_binding("TC1");
+			break;
+		case 2:
+			data->timer = device_get_binding("TC2");
+			break;
+		case 3:
+			data->timer = device_get_binding("TC3");
+			break;
```
I don't know the hardware, but this should not be merged as-is in my opinion.

It looks like the devicetree should be using phandle properties to point to the right timer instances instead. Then you can just use DEVICE_DT_GET to get a pointer to the timer at build time.

I am against any changes to a devicetree binding that imply device_get_binding calls like these. A lot of effort has gone into making device_get_binding unnecessary now that DEVICE_DT_GET can be used to fetch devices from devicetree nodes directly.

mcscholtz replied:
@mbolivar-nordic I agree that is better and that is how I had it originally. The issue I ran into is the case where there is a hardware trigger but it is not a timer. 

Perhaps, instead I can add another KConfig definition that enable/disable all the timer code, which is dependant on DMA being enabled. That way there will be no issue with the case where no timer is provided in the device tree since that code path will be disabled by the KConfig.

The other way is just to require the timer with DMA and not support the other hardware triggers. I will only be using the timer so this is fine for me, but this is less flexible. 

Thoughts?

mbolivar-nordic replied:
Can you explain a bit more what's going on in hardware?

I'm not familiar with how ADC peripheral DMA bindings work, but this is pretty suspicious in general. I would have expected something like this:

https://github.com/torvalds/linux/blob/master/Documentation/devicetree/bindings/iio/adc/st%2Cstm32-adc.yaml#L495

@anangl do you have any general process for how to provide DMA support in the ADC subsystem?

Also cc'ing @erwango since he may be able to shed some light on if the STM32 binding is not relevant here

mcscholtz replied:
@mbolivar-nordic 

Basically the DMA on these devices have different slots for different peripherals. You configure the DMA to use any of the channels and a specific slot that is specific to the peripheral the transfer will be to/from. Instead of assigning it in the device tree this value is determined by the driver based on which ADC it is (ADC0 or ADC1)

You have to manually chose a channel in the device tree because the driver does not currently have a mechanism for providing a free channel.

As for the actual transfer, there is no configuration required on the ADC side, an event will be generated that notifies the DMA there is a sample ready, so long as the correct slot is configured.

The timer is only used here to trigger the ADC conversion, which in-turn notifies the DMA there is a sample waiting to be transferred. There is configuration required for the timer to be used as the trigger source as you can see in the code below:
```
	if (cfg->periph_id == ID_AFEC0) {
		dma_cfg.dma_slot = XDMAC_CC_PERID_AFEC0_Val;
	} else if (cfg->periph_id == ID_AFEC1) {
		dma_cfg.dma_slot = XDMAC_CC_PERID_AFEC1_Val;
	} else {
		__ASSERT_MSG_INFO("AFEC has an invalid peripheral ID!");
		return -EINVAL;
	}
```

The other option is to control the timer outside of the driver. But I found in this case the samples are not always in the correct order because the trigger timing is different each time. When the timer is started/stopped by the driver it is consistent.

I did not implement a way to use DMA without a hardware trigger as it would sort of defeat the purpose and would require extra complexity.

Like I said above I could change the binding to take the phandle of the timer instance instead and add a KConfig that enable/disable the timer code if someone wants to use a different hardware trigger like PWM or the analog comparator.

Right now it looks like this:
```
&afec0
{
	status = "okay";
	compatible = "atmel,sam-afec";
	dma-channel = <0>;
	tc = <0>;
	trigger-source = <1>;
};


&tc0 {
	status = "okay";
	reg-cmr = <0xCC003>;
	reg-rc = <125>;
	channel = <0>;
};
```
But would change to something like this (timer config being the same)

```
&afec0
{
	status = "okay";
	compatible = "atmel,sam-afec";
	dma-channel = <0>;
	tc = <&tc0>;
	trigger-source = <1>;
};
```
There is only a single DMA device that I why I opted not to include it in the devicetree and only include the channel number. Is this what you were referring to?

mcscholtz replied:
@mbolivar-nordic I have changed the way this is done to how I think you would like it. The binding now takes a phandle to the timer and a phandle + channel options to the dma. 

I looked at the sam [i2s](https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/i2s/i2s_sam_ssc.c) driver for inspiration.

Commit:
drivers: adc: atmel_sam: Add DMA support

This patch adds DMA support for the atmel sam s70/e70/v70/v71.
This patch requires that a hardware ADC trigger be configured.
The only hardware trigger that was tested was the timer counter.

Tested on a custom sams70 board.

Signed-off-by: Marius Scholtz <mariuss@ricelectronics.com>

mbolivar-nordic commented on dts/bindings/adc/atmel,sam-afec.yaml:
```diff
@@ -31,5 +31,53 @@ properties:
+      required: false
+      default: 0
+      enum:
+        - 0 # Software driven
+        - 1 # TIOA output of Timer Counter channel 0 for AFEC0
+            # TIOA output of Timer Counter channel 3 for AFEC1
+        - 2 # TIOA output of Timer Counter channel 1 for AFEC0
+            # TIOA output of Timer Counter channel 4 for AFEC1
+        - 3 # TIOA output of Timer Counter channel 2 for AFEC0
+            # TIOA output of Timer Counter channel 5 for AFEC1
+        - 4 # PWM0 event line 0 for AFEC0
+            # PWM1 event line 0 for AFEC1
+        - 5 # PWM0 event line 1 for AFEC0
+            # PWM1 event line 1 for AFEC1
+        - 6 # Analog Comparator
```
If you put this in the description, it will show up in the bindings index:

https://docs.zephyrproject.org/latest/reference/devicetree/bindings.html

I would suggest that because that way users do not have to navigate to the binding file on the file system to look up this useful information you have gathered. But it is not a blocking request.

mbolivar-nordic reviewed (approved):
Thanks for your patience. This looks like a reasonable approach to me from the devicetree side. Hopefully someone who knows this SoC can comment on the driver changes.

nandojve commented on drivers/adc/adc_sam_afec.c:
```diff
@@ -279,6 +365,17 @@ static int adc_sam_init(const struct device *dev)
+		afec->AFEC_MR |= AFEC_MR_TRGEN_EN;
+		afec->AFEC_MR |= AFEC_MR_TRGSEL(cfg->trigger_source);
+	}
+#endif
 	/* Set all channels CM voltage to Vrefp/2 (512). */
```
Add space between `#endif` and `/* comment */`

nandojve commented on drivers/adc/adc_sam_afec.c:
```diff
@@ -296,6 +393,49 @@ static int adc_sam_init(const struct device *dev)
+			.source_address = (uint32_t)&(afec->AFEC_LCDR),
+			.source_addr_adj = DMA_ADDR_ADJ_NO_CHANGE,
+			.dest_addr_adj = DMA_ADDR_ADJ_INCREMENT
+			/* block size and address to be filled when transfer is initiated */
```
Besides Zephyr allows 100 cols, we try keep as 80 in Atmel drivers.

nandojve commented on drivers/adc/adc_sam_afec.c:
```diff
@@ -296,6 +393,49 @@ static int adc_sam_init(const struct device *dev)
+		afec->AFEC_EMR |= AFEC_EMR_TAG;
+#endif
+	}
+	cfg->cfg_func(dev);
+#else
+	cfg->cfg_func(dev);
+#endif
```
Could `cfg->cfg_func(dev);` be only one after `#endif`?

nandojve commented on drivers/adc/adc_sam_afec.c:
```diff
@@ -349,6 +489,26 @@ static void adc_sam_isr(const struct device *dev)
 	}
 }
 
+#define ADC_SAM_GET_TIMER(inst)						\
+	COND_CODE_0(DT_INST_PROP_OR(inst, tc, 0), NULL,			\
+	DEVICE_DT_GET(DT_INST_PHANDLE(inst, tc)))
+
```
It is necessary keep indentation to make statement clear.

```
#define ADC_SAM_GET_TIMER(inst)						\
	COND_CODE_0(DT_INST_PROP_OR(inst, tc, 0),			\
		    NULL,						\
		    DEVICE_DT_GET(DT_INST_PHANDLE(inst, tc)))
```

nandojve commented on drivers/adc/adc_sam_afec.c:
```diff
@@ -349,6 +489,26 @@ static void adc_sam_isr(const struct device *dev)
+	COND_CODE_0(DT_INST_PROP_OR(inst, tc, 0), NULL,			\
+	DEVICE_DT_GET(DT_INST_PHANDLE(inst, tc)))
+
+#define ADC_SAM_GET_DMA(inst)						\
+	COND_CODE_0(DT_INST_PROP_OR(inst, dmas, 0),			\
+	NULL,								\
+	(.dma = DEVICE_DT_GET(DT_INST_DMAS_CTLR_BY_IDX(inst, 0)),	\
+	.dma_channel = DT_INST_DMAS_CELL_BY_IDX(inst, 0, channel),	\
+	.dma_slot = DT_INST_DMAS_CELL_BY_IDX(inst, 0, perid)))		\
+
+#ifdef CONFIG_ADC_SAM_AFEC_DMA
+#define ADC_SAM_DMA_CONFIG(n)						\
+	ADC_SAM_GET_DMA(n),						\
+	.trigger_source = DT_INST_PROP_OR(n, trigger_source, 0),	\
+	.tc = ADC_SAM_GET_TIMER(n),
+#else
```
Add indentation.

nandojve commented on drivers/adc/adc_sam_afec.c:
```diff
@@ -154,6 +210,19 @@ static void adc_context_start_sampling(struct adc_context *ctx)
+		data->dma_blk.block_size = ctx->sequence.buffer_size;
+		data->dma_blk.dest_address = (uint32_t)data->buffer;
+		dma_config(cfg->dma, cfg->dma_channel, &data->dma_cfg);
+	}
+
+#endif
```
```diff
	}
-
#endif
+
	adc_sam_start_conversion(data->dev);
```

nandojve commented on drivers/adc/adc_sam_afec.c:
```diff
@@ -139,7 +151,51 @@ static void adc_sam_start_conversion(const struct device *dev)
+	}
+#else
+	adc_sam_start_sw_conversion(dev);
+#endif
```
```diff
#endif
+
	/* Start the conversions. */
```

nandojve reviewed (changes requested):
Thank you for this contribution @mcscholtz !

Few comments in the PR. What is important is make sure CI can test config with DMA enable. Please, review "tests/drivers/adc_dma" to see if it is possible make it generic to use with other platforms like 'sam_e70_xplained'.

You can create many config as requested to tune any test:
https://github.com/zephyrproject-rtos/zephyr/blob/5fd079bd68f381e2907d54dacc040a48c03b64ad/samples/net/sockets/echo_client/sample.yaml#L20-L34
